### PR TITLE
JIT `wigner()`

### PR DESCRIPTION
**Pull request by abocquet**
Clenshaw only. Not the cleanest but I speeds up plot mosaic a bit. On a CPU:
image

For later PRs:
- [ ] FFT method
- [ ] implement a batching parameter to have a tradeoff between memory efficiency and speed

### CODE REVIEW

File: dynamiqs/utils/wigners.py [hunk 1]
```diff
@@ -45,17 +45,29 @@ def wigner(
...
     xvec = jnp.linspace(-xmax, xmax, npixels, dtype=rdtype)
     yvec = jnp.linspace(-ymax, ymax, npixels, dtype=rdtype)
 
     if method == 'clenshaw':
         state = todm(state)
-        w = _wigner_clenshaw(state, xvec, yvec, g)
+        expanded = False
+        if state.ndim == 2:
+            state = state[None, ...]
+            expanded = True
```
Comment (pierreguilmin) [inline dynamiqs/utils/wigners.py:57]: Do we really need to do this every time we want to batch over a potentially non-existing dimension?  (especially having to remember `expanded` is a bit annoying) Is there a shorter way/trick/one liner/utility function/context manager we could use?

File: dynamiqs/utils/wigners.py [hunk 2]
```diff
@@ -45,17 +45,29 @@ def wigner(
     """
     state = jnp.asarray(state)
```
Comment (pierreguilmin) [inline dynamiqs/utils/wigners.py:46]: Ideally we can vmap `wigner()` directly once we fix fft.

File: dynamiqs/utils/wigners.py [hunk 3]
```diff
@@ -45,17 +45,29 @@ def wigner(
...
     if method == 'clenshaw':
         state = todm(state)
-        w = _wigner_clenshaw(state, xvec, yvec, g)
+        expanded = False
+        if state.ndim == 2:
+            state = state[None, ...]
+            expanded = True
+
+        n = state.shape[-1]
+        w = jax.vmap(_wigner_clenshaw, in_axes=(0, None, None, None, None))(
```
Comment (pierreguilmin) [inline dynamiqs/utils/wigners.py:60]:
```suggestion
        w = jax.vmap(_wigner_clenshaw, in_axes=(-3, None, None, None, None))(
```
Can you try and see if it works on doubly-batched arrays? (e.g. `(8, 5, 16, 16)`)

**Review by pierreguilmin — APPROVED** [review]
Great 😍

**Comment by abocquet** [comment]
My bad the screenshot was wrong
image

File: dynamiqs/utils/wigners.py (hunk 1 above)
Comment (abocquet) [reply dynamiqs/utils/wigners.py:57]: I'm open to any idea

File: dynamiqs/utils/wigners.py (hunk 3 above)
Comment (abocquet) [reply dynamiqs/utils/wigners.py:60]: I'm not sure what you want to do with `-3` 🤔 
It would work with a double vmap

Maybe jitting wigner direclty would be better, I'm going to see what we can do

Comment (pierreguilmin) [reply dynamiqs/utils/wigners.py:60]: I thought "iterate over the third axis starting from last of the third element", to support `(..., n, n)` and not just `(b, n, n)`, but it's maybe not the correct way to do. We want the function to work with arbitrarily batched input.

File: dynamiqs/utils/wigners.py (hunk 1 above)
Comment (pierreguilmin) [reply dynamiqs/utils/wigners.py:57]: I'll think about it 😊

**Comment by abocquet** [comment]
You were right, it's much more natural to batch `wigner`

**Comment by pierreguilmin** [comment]
I'm confused, it's jitted but it's not batched?

**Comment by abocquet** [comment]
No, if you want to batch you'd have to use `vmap`

**Comment by pierreguilmin** [comment]
Isn't the goal of the PR to make wigner batch-compatible to use it in `plot_wigners_mosaic`?

**Review by pierreguilmin — APPROVED** [review]

File: dynamiqs/utils/wigners.py [hunk 4]
```diff
@@ -15,6 +15,8 @@
 __all__ = ['wigner']
 
 
+@ft.partial(jnp.vectorize, signature='(n,m)->(k),(l),(k,l)', excluded={1, 2, 3})
+@ft.partial(jax.jit, static_argnums=(3, 4))
```
Comment (pierreguilmin) [inline dynamiqs/utils/wigners.py:19]: Should we reverse both? I don't know if vectorize jits by default.

**Review by pierreguilmin — APPROVED** [review]
Beautiful!

File: dynamiqs/utils/wigners.py (hunk 4 above)
Comment (abocquet) [reply dynamiqs/utils/wigners.py:19]: It's sensitively equivalent

File: dynamiqs/utils/wigners.py [hunk 5]
```diff
@@ -73,6 +72,7 @@ def wigner(
     return xvec, yvec, w
 
 
+@ft.partial(jax.jit, static_argnums=(4,))
```
[the remark this answers is not in the document]
Comment (abocquet) [reply dynamiqs/utils/wigners.py]: Yes

**Review by pierreguilmin — APPROVED** [review]
